---
title: "software_comparison"
author: "Rachel Xu"
date: "1/3/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/rx32940/Dropbox/5.Rachel-projects/Metagenomic_Analysis/revision/Metagenomics_tools")

software.xlsx <- openxlsx::createWorkbook()
```
# load analyses output of all software
```{r}
input <- "1.output"

blastn_phylo <- read_blastn("2.report/blastn") # 3890 taxa
diamond_phylo <- read_diamond("2.report/diamond") # 1378 taxa
kraken2_phylo <- read_kraken2(file.path(input, "kraken2.std.biom")) # 3074 taxa
centrifuge_phylo <- read_kraken2(file.path(input, "centrifuge.biom")) # 4571 taxa
kaiju_phylo <- read_kraken2(file.path(input, "kaiju.cus.biom")) # 4816 taxa
bracken_phylo <- read_kraken2(file.path(input, "bracken.std.biom")) # 505 taxa
metaphlan3_phylo <- read_metaphlan3(file.path(input, "metaphlan3_all_samples.txt")) # 18 taxa
clark_phylo <- read_clark("2.report/clark/") # 1628 taxa
clarks_phylo <- read_clark_s("2.report/clark-s/") # 1376 taxa

software_list <- list(blastn=blastn_phylo, diamond = diamond_phylo,kraken2=kraken2_phylo,bracken=bracken_phylo,centrifuge = centrifuge_phylo,clark = clark_phylo,clarks = clarks_phylo,metaphlan3 = metaphlan3_phylo,kaiju =kaiju_phylo)

softwares <- c(names(software_list))
samples <- factor(c("R22.K","R26.K","R27.K","R28.K","R22.L","R26.L","R27.L","R28.L","R22.S","R26.S","R27.S","R28.S" ), levels = c("R22.K","R26.K","R27.K", "R28.K","R22.L","R26.L","R27.L","R28.L","R22.S","R26.S","R27.S","R28.S" ))
samples_df <- metadata %>% rownames_to_column(var="samples")


```


# combine classification of different softwares
- combine number of reads classified by each DB in to the same data frame 
- add taxonomy ranks to each taxID
- write to file
```{r}

# combine otu tables of all databases
#### need to first run lines in the loop before decale combined db
all_df_count <- data.frame(rowname =otu_df$rowname)
for (sf in softwares){
  soft_phylo <- paste0(sf, "_phylo")
  otu_df <- as.data.frame(otu_table(get(soft_phylo)))  %>% dplyr::rename_with(~paste(.x, eval(sf), sep = ".")) %>% rownames_to_column()
  all_df_count <-full_join(otu_df,all_df_count, by = "rowname")
}

all_df_count[is.na(all_df_count)] <- 0

# combine taxonomy ranks tables of all databases
all_df_taxa <- data.frame(rowname=as.character(),Domain=as.character(), Phylum=as.character(), Class=as.character(), Order=as.character(), Family=as.character(), Genus=as.character(),Species=as.character())
for (sf in softwares){
  soft_phylo <- paste0(sf, "_phylo")
  tax_df <- as.data.frame(tax_table(get(soft_phylo))) %>% rownames_to_column()
  all_df_taxa <- full_join(all_df_taxa, tax_df, by = c("rowname","Domain", "Phylum", "Class", "Order", "Family", "Genus"))
}

# merge species names from different software
all_df_taxa$Species <- ifelse(!is.na(all_df_taxa$Species.y.y.y.y.y), all_df_taxa$Species.y.y.y.y.y, ifelse(
    !is.na(all_df_taxa$Species.y.y.y), all_df_taxa$Species.y.y.y, ifelse(
      !is.na(all_df_taxa$Species.y.y), all_df_taxa$Species.y.y, ifelse(
        !is.na(all_df_taxa$Species.x.x.x), all_df_taxa$Species.x.x.x, ifelse(
          !is.na(all_df_taxa$Species.x.x.x.x.x), all_df_taxa$Species.x.x.x.x.x, ifelse(
            !is.na(all_df_taxa$Species.x.x.x.x), all_df_taxa$Species.x.x.x.x, ifelse(
              !is.na(all_df_taxa$Species.y.y.y.y), all_df_taxa$Species.y.y.y.y, ifelse(
                !is.na(all_df_taxa$Species.x.x), all_df_taxa$Species.x.x, ifelse(
                  !is.na(all_df_taxa$Species.y), all_df_taxa$Species.y, ifelse(
                    !is.na(all_df_taxa$Species.x), all_df_taxa$Species.x, NA
                  )
                )
              )
            )
          )
        )
    )
  )
))

all_df_taxa_merge <- all_df_taxa %>% dplyr::select(!starts_with("Species.x") & !starts_with("Species.y")) # remove conflicting species name

all_df_taxa_unique <- all_df_taxa_merge %>%mutate(n_char = sapply(Species, nchar))  %>% group_by(rowname, Domain, Phylum, Class, Order, Family, Genus) %>% dplyr::slice(which.min(n_char)) %>% dplyr:: select(!n_char) %>% as.data.frame() %>% dplyr::distinct(rowname, .keep_all = TRUE) # if there is duplicate in species name, select the one with the shortest name

# combine taxonomy ranks with count of the ranks
software_df <- left_join(all_df_count, all_df_taxa_unique , by = "rowname") %>% dplyr::rename(TaxID = rowname)
# colnames(software_df)

# reorder columns of the df
software_df <- software_df[c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species","TaxID",paste0(samples, ".blastn"),paste0(samples, ".diamond"), paste0(samples, ".kraken2"), paste0(samples, ".bracken"), paste0(samples, ".centrifuge"), paste0(samples, ".clark"), paste0(samples, ".clarks"), paste0(samples, ".metaphlan3"),  paste0(samples, ".kaiju"))]

# add_sheet(software.xlsx,"1.software.full", software_df)

```

# identify Leptospira
```{r}


lepto.xlsx <- openxlsx::createWorkbook()
lepto_full <- software_df %>% subset(Genus == "g__Leptospira") %>% pivot_longer(cols = colnames(software_df)[!colnames(software_df) %in%c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species","TaxID")], names_to = "samplessoftware", values_to = "count") %>% tidyr::separate(samplessoftware, c("samples", "Tissue","software"), sep="[.]+") %>% unite("sample",c("samples", "Tissue"), sep=".")

unique_sp_count <- lepto_full %>% group_by(sample, software) %>% subset(count !=0) %>% summarise(n_species = n_distinct(TaxID))
unique_sp <- lepto_full %>% group_by(sample, software) %>% subset(count !=0) %>% distinct(TaxID, .keep_all = TRUE)
lepto_count_summary <- lepto_full %>% group_by(sample, software) %>% subset(count !=0) %>% summarise(total=sum(count)) 
lepto_count_summary %>% group_by(sample) %>% summarise(mean=mean(total), sd=sd(total))
lepto_present_matrix <- unique_sp_count %>% dplyr::select(-n_species)%>%  mutate(present = "x")%>%pivot_wider(names_from = sample, values_from = present)
lepto_present_matrix[is.na(lepto_present_matrix)] <- ""

lepto_count_summary %>% group_by(sample) %>% summarise(mean=mean(total), sd=sd(total))
lepto_present_matrix
# add_sheet(lepto.xlsx,"1.software.identified.lepto", unique_sp)
# add_sheet(lepto.xlsx,"2.unique.Species.count", unique_sp_count)
# add_sheet(lepto.xlsx,"present_absent", lepto_present_matrix)
# add_sheet(lepto.xlsx,"3.lepto.count.summary", lepto_count_summary)

# openxlsx::saveWorkbook(lepto.xlsx, file = "4.tables/III.lepto.diagnostic.xlsx", overwrite = TRUE)
```

# change the taxa table with the tax table merged from all softwares
- uniform species taxa
```{r}
all_df_uniqueTax <- as.matrix(all_df_taxa_unique %>% column_to_rownames(var="rowname") )


tax_table(blastn_phylo) <- all_df_uniqueTax
tax_table(diamond_phylo) <- all_df_uniqueTax
tax_table(kraken2_phylo) <- all_df_uniqueTax
tax_table(bracken_phylo) <- all_df_uniqueTax
tax_table(centrifuge_phylo) <- all_df_uniqueTax
tax_table(clark_phylo) <- all_df_uniqueTax
tax_table(clarks_phylo) <- all_df_uniqueTax
tax_table(metaphlan3_phylo) <- all_df_uniqueTax
tax_table(kaiju_phylo) <- all_df_uniqueTax


##### check if all phylo object has the same species taxa
# intersect(rownames(tax_table(blastn_phylo)), rownames(tax_table(kaiju_phylo)))
# 
# tax_table(blastn_phylo)[rownames(tax_table(blastn_phylo)) == "173",]
# tax_table(kaiju_phylo)[rownames(tax_table(kaiju_phylo)) == "173",]
```


## get domain level read counts for each software
```{r}

software_domain_count <- data.frame(taxa = character(), Domain = character(), samples = character(), count=numeric(), Tissue=character(), Subject=character(), software=character())
for(i in 1:length(software_list)){
  otu <- data.frame(otu_table(tax_glom(unname(software_list[i])[[1]]), "Domain")) %>% rownames_to_column(var="taxa")
  lineage_df <- data.frame(tax_table(unname(software_list[i])[[1]])) %>% rownames_to_column(var="taxa")
   domain_df <- left_join(otu, lineage_df, by="taxa") %>% pivot_longer(cols =samples, names_to = "samples", values_to = "count")
  domain_anot_df <- left_join(domain_df,samples_df, by="samples" ) %>% mutate(Domain = sapply(Domain, function(x){
    unlist(strsplit(x, "__"))[2]
  })) %>%  mutate(software=names(software_list[i]))
  
  software_domain_count <- rbind(software_domain_count,domain_anot_df)

}

numTaxa <- as.data.frame(unlist(lapply(software_list,ntaxa))) %>% rownames_to_column()
colnames(numTaxa) <- c("software", "ntaxa")
software_summary <- software_domain_count %>% group_by(samples, software) %>% summarise(total=sum(count)) %>% pivot_wider(names_from = software, values_from = total) # summary of count per sample
software_summary_mean <-software_domain_count %>% group_by(samples, software) %>% summarise(total=sum(count))%>% group_by(software) %>% summarise(mean=mean(total), sd=sd(total))
software_summary_mean <-left_join(software_summary_mean,numTaxa, by ="software")

# add_sheet(software.xlsx,"2.software.samples.summary", as.data.frame(software_summary))
# add_sheet(software.xlsx,"2.software.summary", software_summary_mean)

```
### plot number of reads classified under each domain
```{r}
# software_domain_count$samples <- factor(software_domain_count$samples, levels=samples)
# software_domain_count$software<- factor(software_domain_count$software, levels = softwares)
shapes <- c(16:18, 6:14)

software_domain_subset <- software_domain_count %>% subset(!Domain %in% c("unknown"))
software_domain_subset$samples <- factor(software_domain_subset$samples, levels=unique(software_domain_subset$samples))
software_domain_subset$software<- factor(software_domain_subset$software, levels = unique(software_domain_subset$software))
unique(software_domain_subset$Domain)
software.domain.stat <- software_domain_subset %>%
  group_by(Domain) %>%
  wilcox_test(count~software,p.adjust.method ="holm", paired = TRUE) %>%
  add_significance()%>%
  add_xy_position(x = "software", dodge = 0.1, scales = "free") %>%
  add_y_position(fun="max", step.increase = 0.08, scales = "free")

# only keep stat with significant p value, change the y position
software.domain.stat[software.domain.stat["Domain"] == "Bacteria" & software.domain.stat["p.adj.signif"] != "ns", "y.position"] <- seq(120000, 120000+ 12000 * 6, 12000) # minimum y position for the domain, n-1 incremention where n is comparison that is significant
software.domain.stat[software.domain.stat["Domain"] == "Archaea" & software.domain.stat["p.adj.signif"] != "ns", "y.position"] <- seq(500, 500 + 150 * 4, 150)
software.domain.stat[software.domain.stat["Domain"] == "Viruses" & software.domain.stat["p.adj.signif"] != "ns", "y.position"] <- seq(1000, 1000 + 150 * 6, 150)

software_domain_subset$Domain <- factor(software_domain_subset$Domain, levels=c("Eukaryota", "Bacteria", "Viruses", "Archaea"))
p <- ggplot(software_domain_subset, aes(x=software, y=count))+
  geom_boxplot(outlier.shape = NA,color="gray")+
  scale_shape_manual(values=shapes)+
  geom_point(aes(shape=samples),size=2, position = position_jitter(0.3))+
  # stat_pvalue_manual(software.domain.stat, label = "p.adj.signif", hide.ns = TRUE,size =5, bracket.shorten =0 ,bracket.size = 0.3, tip.length = 0.01)+
  facet_wrap(~Domain, scales = "free",ncol=4)+
  theme_classic()+
  theme(text=element_text(size=18), title=element_text(size=17),axis.text.x = element_text(angle=90,hjust=0.95,vjust=0.2),tagger.panel.tag.background = element_rect(linetype = 0,fill = alpha("white", 0)))+
  tag_facets()+
  scale_y_continuous(label=comma)+
  labs(x="Software", y="Domain Level Read Count")


p

software_domain_subset %>% group_by(software, Domain, samples) %>%  summarise(mean=mean(count)) %>% group_by(software, Domain) %>%  summarise(mean=mean(mean)) %>% subset(Domain == "Viruses")
# add_sheet(software.xlsx,"3.software.domain.stats", software.domain.stat %>% dplyr::select(Domain, group1, group2, p, p.adj, p.adj.signif))
# ggsave("5.figures/II.software.comparison.domain.counts.pdf", p, width = 10, height = 6)

```

# microbiom analysis
- host taxa (Eukaryota) were **removed** from the downstream analysis
- also, taxa for all software are aggregating at Species level to 
```{r}
software_w_eukary <- c("blastn", "diamond", "centrifuge", "clark", "bracken", "kraken2")
for(software in names(software_list)){
  print(eval(software))
  software_obj <- paste(software,"phylo", sep = "_")
  if(software %in% software_w_eukary){
    eukary_taxa <- rownames(tax_table(get(software_obj))[tax_table(get(software_obj))[,"Domain"] == "d__Eukaryota",])
    eurkary_phylo <- pop_taxa(get(software_obj), eukary_taxa)
    print(paste(eval(software_obj),"with eukaryota"))
  }else{
    eurkary_phylo <-get(software_obj)
    print(paste(eval(software_obj),"without eukaryota"))
  }
  
  
  eukary_sp <- tax_glom(eurkary_phylo, "Species")
  assign(paste(software,"sp", sep = "_"), eukary_sp)
}

software_species <- list(blastn=blastn_sp, diamond = diamond_sp,kraken2=kraken2_sp,bracken=bracken_sp,centrifuge = centrifuge_sp,clark = clark_sp,clarks = clarks_sp,metaphlan3 = metaphlan3_sp,kaiju =kaiju_sp)

# lapply(software_species, ntaxa)


```
# characterize samples at the species level

### species shared by all softwares
```{r}
library(UpSetR)
library(ComplexHeatmap)
sp_software_list <- unlist(lapply(software_species, ntaxa))

unique_sp <- function(phylo_sp){
  unique(data.frame(tax_table(phylo_sp)) %>% dplyr::select(Species) %>% unlist())
}

unique_species_list <- lapply(software_species, unique_sp)
# unique_species_list

intersect_sp <- Reduce(intersect,unique_species_list) # 9 intersection across all software
# s1 <- data.frame(tax_table(kaiju_sp))
# s1 %>% rownames_to_column() %>% subset(Species %in% intersect_sp)
# use complexheatmap to make intersection
# blastn, diamond, kraken2, bracken, centrifuge, clark, clarks, metaphlan3, kaiju
spm <- make_comb_mat(unique_species_list, mode="intersect")

# length(intersect(unique_species_list$centrifuge,unique_species_list$kraken2))
species_intersectin_set <- data.frame(size= comb_size(spm)[order(comb_size(spm), decreasing = TRUE)])
# Sets named by presence (1)/absence (0) of the software in the intersection with the order of: Blastn, Diamond, Kraken2, Bracken, Centrifuge, CLARK, CLARK-s, Metaphlan3, and Kaiju	
# add_sheet(software.xlsx,"4.species.overlap.size", species_intersectin_set %>% rownames_to_column())
# UpSet(spm, pt_size=unit(2, "mm"))

```

### species with at least 10% reads from each samples identified from each software
```{r}

phylo_sp <- software_species[[1]]
tenPer_sp <- function(phylo_sp){
tenPer_df <- as.data.frame(otu_table(phylo_sp)) %>% rownames_to_column()%>%pivot_longer(names_to="samples", values_to="count", cols =samples)%>% group_by(samples) %>% mutate(total = sum(count)) %>% mutate(tenPercent = total * 0.1) %>% subset(count >= tenPercent) %>% arrange(desc(count), .by_group = TRUE)
unique(tenPer_df$rowname)
}


tenPer_sp_software_list <- lapply(software_species, tenPer_sp)
table(unlist(tenPer_sp_software_list))
tenPer_sp_software_list_count <- lapply(tenPer_sp_software_list , length)

intersect_tenPer <- Reduce(intersect, tenPer_sp_software_list[names(tenPer_sp_software_list)] )
# s1 <- data.frame(tax_table(kaiju_sp))
s1 %>% rownames_to_column() %>% subset(rowname %in% c(173,807))

mtp <- make_comb_mat(tenPer_sp_software_list, mode = "intersect")

```

## estimate alpha diversity obtained from the classification of each software
```{r}

software_species <- list(blastn=blastn_sp, diamond =diamond_sp, kraken2=kraken2_sp,bracken=bracken_sp,centrifuge= centrifuge_sp,clark = clark_sp,clarks = clarks_sp,metaphlan = metaphlan3_sp,kaiju =kaiju_sp) # 2295, 203,2749, 504, 4191,1627,1376,18, 4128

software.alpha <- data.frame(samples = character(), Observed=numeric(), Shannon = numeric(), Simpson = numeric(), database=character())
for(i in 1:length(software_species)){
  
 software.alpha <- rbind(software.alpha, estimate_richness(unname(software_species[i])[[1]], measures = c("Observed","shannon", "Simpson")) %>% rownames_to_column(var="samples") %>% mutate(software = names(software_list[i])))
}


 software.alpha.longer <-  software.alpha %>% pivot_longer(c("Observed","Shannon", "Simpson"), names_to = "alpha_index", values_to = "values")


```



```{r}

software.alpha.longer$samples <- factor(software.alpha.longer$samples, levels=samples)
software.alpha.longer$software <- factor(software.alpha.longer$software, levels = softwares)
shapes <- c(16:18, 6:14)

alpha.stat.soft <- software.alpha.longer %>%
  group_by(alpha_index) %>%
  wilcox_test(values~software,p.adjust.method ="holm", paired = TRUE) %>%
  add_significance()%>%
  add_xy_position(x = "software", dodge = 0.1, scales = "free") %>%
  add_y_position(fun="max", step.increase = 0.2, scales = "free")

alpha.stat.soft[alpha.stat.soft["alpha_index"] == "Observed" & alpha.stat.soft["p.adj.signif"] != "ns", "y.position"] <- seq(3483, 3483+ (500 * 29), 500)
alpha.stat.soft[alpha.stat.soft["alpha_index"] == "Shannon" & alpha.stat.soft["p.adj.signif"] != "ns", "y.position"] <- seq(7, 7+ 0.6 * 22, 0.6)
alpha.stat.soft[alpha.stat.soft["alpha_index"] == "Simpson" & alpha.stat.soft["p.adj.signif"] != "ns", "y.position"] <- seq(1.1, 1.1+ 0.05 * 6, 0.05)

p <- ggplot(software.alpha.longer, aes(x=software, y=values))+
  geom_boxplot(outlier.shape = NA, color = "gray")+
  scale_shape_manual(values=shapes)+
  geom_point(aes(shape=samples),size=3, position = position_jitter(0.2))+
  facet_wrap(~alpha_index, scales = "free_y")+
  # stat_pvalue_manual(alpha.stat.soft, label = "p.adj.signif", hide.ns = TRUE,size=5, bracket.shorten =0 ,bracket.size = 0.3, tip.length = 0.01)+
  theme_classic()+
  theme(text=element_text(size=20), title=element_text(size=17),axis.text.x = element_text(angle=90,hjust=0.95,vjust=0.2),tagger.panel.tag.background = element_rect(linetype = 0,fill = alpha("white", 0)))+
  tag_facets()+
  scale_y_continuous(label=comma)+
  labs(x="Software", y="Species Level Alpha Diversity")
p


# add_sheet(software.xlsx,"5.alpha.stats", alpha.stat.soft %>% dplyr::select(alpha_index, group1, group2, p, p.adj, p.adj.signif))
# ggsave("5.figures/II.software.comparison.alpha.indices.pdf", p,width = 10,height = 7)

```

# beta diversity

### function to plot beta heatmap
```{r}
software <- "blastn"
plot_beta_heatmap <- function(software){
phyloseq_obj <- paste0(software, "_sp")
annot_colors<- list(Subject= c(R22 = "#0b7fab", R26="#e9723d", R27="#f4d75e", R28="#7c7b89"), Tissue=c(Kidney = "#c33124", Lung="#a1dffb", Spleen="#e8a628"))
beta_dist_db <- as.matrix(vegdist(t(otu_table(get(phyloseq_obj))), index="bray"))
beta_dist_db[is.na(beta_dist_db)] <- 1 # 1 in bray curtis index means not related
p <- pheatmap::pheatmap(beta_dist_db, col = colorRampPalette(c("red", "#FFA701","lightyellow"))(50), annotation = metadata, annotation_colors = annot_colors,silent = TRUE, legend = FALSE, annotation_legend = FALSE, treeheight_row = 0, treeheight_col = 30,main = software)
# ggsave("5.figures/legend.pdf",p, width = 12, height = 12)
}

```

### generate beta hap map for all kraken2 databses
```{r}
software_species <- list("blastn", "diamond","kraken2","bracken","centrifuge","clark","clarks","metaphlan3","kaiju")

software_beta_heatmaps <- lapply(software_species, plot_beta_heatmap)


combined_software_heatmap <- grid.arrange(grid.arrange(),grobs = list(software_beta_heatmaps[[1]][[4]], software_beta_heatmaps[[2]][[4]], software_beta_heatmaps[[3]][[4]], software_beta_heatmaps[[4]][[4]], software_beta_heatmaps[[5]][[4]], software_beta_heatmaps[[6]][[4]], software_beta_heatmaps[[7]][[4]],software_beta_heatmaps[[8]][[4]],software_beta_heatmaps[[9]][[4]]),ncol =3) # fourth item of every pheatmap object is the plot

# ggsave("5.figures/II.software.comparison.beta.heatmap.pdf", combined_software_heatmap,width = 12, height = 12)

```
### beta diversity statistics
```{r}
# must run one loop first
beta_dist_stat <- data.frame(comparison = beta_dist_df_combined$comparison )

for (software in softwares){
  soft_phylo <- paste(software, "phylo", sep="_") 
  beta_dist <-as.matrix(vegdist(t(otu_table(get(soft_phylo))), index="bray")) 
  beta_dist[is.na(beta_dist)] <- 1
  beta_dist_upper <- melt(beta_dist)[melt(upper.tri(beta_dist))$value,]
  ifelse(software == "metaphlan3",beta_dist_df_combined <- beta_dist_upper %>% mutate(comparison = paste(Var1, Var2, sep="-")) %>% dplyr::select(-c(Var1, Var2)) , beta_dist_df_combined <- beta_dist_upper %>% mutate(comparison = paste(Var2, Var1, sep="-")) %>% dplyr::select(-c(Var1, Var2)) )
  colnames(beta_dist_df_combined) <-  c(eval(software), "comparison")
  beta_dist_df_combined$comparison <- sort(beta_dist_df_combined$comparison)
  beta_dist_stat <- left_join(beta_dist_stat, beta_dist_df_combined, by="comparison")
}


beta_dist_stat_longer <- beta_dist_stat %>% pivot_longer(cols = softwares, names_to = "software", values_to = "bray_curtis" )

# if the between samples diversity different when classified by different databases
beta_stat <- beta_dist_stat_longer %>%
  wilcox_test(`bray_curtis`~software, paired = TRUE, p.adjust.method = "holm") 



# add_sheet(software.xlsx,"6.beta.stats", beta_stat %>% dplyr::select(.y., group1, group2, p, p.adj, p.adj.signif))
# write.xlsx(beta_stat_db,"4.tables/I.kraken2_db_comparison.xlsx",sheetName = "4.DB.beta.stat", append=TRUE)
```


# significantly abudnant taxa
```{r}
library(DESeq2)
# software <- "kaiju"
ref_tissue <- "Kidney"
contrast_t <- "Spleen"
# function to do differential abudance analyses
# software <- softwares[1]
diff_abund_test <- function(software){
  software_obj <- paste(software, "sp", sep="_")
  KL_subset <- subset_samples(get(software_obj), Tissue %in%  c(ref_tissue , contrast_t))
  sample_data(KL_subset)$Tissue <- relevel(factor(unique(sample_data(KL_subset)$Tissue)), ref_tissue)
  
  deseq2_obj <- phyloseq_to_deseq2(KL_subset, ~ Tissue)
  deseq_out <- DESeq(deseq2_obj, test="Wald", sfType="poscounts")
  
  # resultsNames(deseq_out)
  # plotDispEsts(deseq_out)
  
  LK_sigTax <- results(deseq_out, cooksCutoff = FALSE, contrast = c("Tissue", ref_tissue,contrast_t))
  sign_taxa <- data.frame(LK_sigTax[which(LK_sigTax$padj < 0.05),]) %>% rownames_to_column()
  KL_taxa <- data.frame(tax_table(KL_subset))%>% rownames_to_column()
  sign_taxa_lineage <- left_join(sign_taxa, KL_taxa, by="rowname") %>% mutate(software = software)
  sign_taxa_lineage
}


# sig_tax_df <- DA_df[[1]]
summarize_DA <- function(sig_tax_df){
  software <- sig_tax_df$software[1]
  neg <- sig_tax_df %>% subset(log2FoldChange <0) %>% dplyr::select(Domain,Phylum, Genus, rowname) %>% group_by(Domain,Phylum, Genus) %>% summarise(negative =n())
  pos <- sig_tax_df %>% subset(log2FoldChange >0) %>% dplyr::select(Domain,Phylum,Genus,rowname) %>% group_by(Domain,Phylum, Genus) %>% summarise(positive =n())
  summary_df <- full_join(neg, pos, by=c("Domain","Phylum","Genus")) %>% pivot_longer(names_to = "direction", values_to = "n_species", cols = c("negative","positive")) %>% mutate(software =software)
  summary_df[is.na(summary_df)] <- 0
  summary_df
}



```


```{r}

# diff_abund_test(software="kaiju")
DA_df <- lapply( softwares[softwares != "metaphlan3"], diff_abund_test) 
# number of diffAbundant species  Lung vs. Kidney:172, 10, 141, 139, 596, 66,51, 549
# number of diffAbundant species Lung vs. Spleen: 169, 44, 159, 171, 457, 86, 57, 484
# number of diffAbundant species Kidney vs. Spleen: 57,6,42,41, 25, 47,38, 38

combined_DA_species <- do.call(rbind, DA_df )

i <- 1
sp_list <- list()
for(i in 1:8){
  sp_list <- append(sp_list,list(DA_df[[i]]$Species))
  i <- i + 1
}
names(sp_list) <- softwares[softwares != "metaphlan3"]
Reduce(intersect, sp_list)
DA_df[[8]][DA_df[[8]]["Species"] == "s__neurolyticum", ]
# add_sheet(software.xlsx,"7.diffAbd_LvK", combined_DA_species %>% dplyr::rename(TaxID = rowname))
# add_sheet(software.xlsx,"8.diffAbd_KvS", combined_DA_species %>% dplyr::rename(TaxID = rowname))
# add_sheet(software.xlsx,"9.diffAbd_LvS", combined_DA_species %>% dplyr::rename(TaxID = rowname))
DA_summary <- lapply(DA_df, summarize_DA)

combined_DA_summary <- do.call(rbind, DA_summary)

combined_DA_summary <- combined_DA_summary %>% mutate(direction = sapply(direction, function(x){
  ifelse(x=="positive", eval(contrast_t), eval(ref_tissue))
}))

combined_DA_summary$Phylum<-  mapply(function(x, y){
  ifelse(x == "" & y != "", paste0("p__", unlist(strsplit(y, "__", fixed = TRUE))[2]), ifelse(
    x == "" & y == "", "p__unknown", x))
},combined_DA_summary$Phylum, combined_DA_summary$Genus)

combined_DA_summary$Genus <- mapply(function(x, y){
  ifelse(x == "" & y != "", paste0("g__", unlist(strsplit(y, "__", fixed = TRUE))[2]), ifelse(
    x == "" & y == "", "c__unknown", x))
}, combined_DA_summary$Genus,combined_DA_summary$Phylum)

# DA_df[[2]]
```


```{r}

manual_color <-c("#A0C4FF","#80B918","#FFC6FF","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#9BF6FF","#BDB2FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#FEE440","#54478C")

combined_DA_summary %>% subset(software =="kaiju")
t1 <- combined_DA_summary %>% subset(Genus != "p__unknown")%>% subset(n_species != 0) %>% dplyr::select(Genus, software, direction)
p1 <- ggplot(combined_DA_summary %>% subset(Genus != "p__unknown")%>% subset(n_species != 0) %>% dplyr::select(Genus, software, direction) , aes(x=software))+
  geom_bar(aes(fill=direction))+
  geom_text(stat='count', aes(label=..count..), hjust = 1)+
  coord_flip()+
  scale_y_reverse()+
  theme_classic()+
  theme(text = element_text(size=16),axis.text.x = element_text(angle= 90), legend.position = "none", axis.text.y = element_blank(), axis.title.y = element_blank(),axis.title.x = element_text(vjust =-0.5),)+
  labs(y="Differentially Abundant Genera")+
  scale_x_discrete(position = "top")+
  coord_flip()+
scale_fill_manual(values= manual_color)
  # geom_text(aes(label=sum(n_species)), stat="identity")
p1 

unique_diffabd_phyla <- combined_DA_summary %>% subset(Phylum != "p__unknown")%>% subset(n_species != 0) %>% group_by(software, direction) %>% summarise(n_phylum = n_distinct(Phylum))


p2 <- ggplot(unique_diffabd_phyla, aes(x= software, y= n_phylum)) +
  geom_bar(stat="identity", aes(fill=direction))+
  theme_classic()+
   stat_summary(
    aes(label = stat(y)), fun.y = 'sum', geom = 'text', hjust = -0.1
  ) +
  theme(text = element_text(size=16),axis.text.x = element_text(angle= 90),axis.text.y = element_text(hjust=0.5), axis.title.y =element_blank() )+
  labs(y="Differentially Abundant Phyla")+
  coord_flip()+
scale_fill_manual(values= manual_color)
p2
plot <- grid.arrange(cbind(ggplotGrob(p1), ggplotGrob(p2), size = "last"))

unique_diffabd_phyla

# ggsave("5.figures/III. software_diffAbdundant_bar_LvK.pdf", plot, width =15)
# ggsave("5.figures/III. software_diffAbdundant_bar_LvS.pdf", plot, width =15)
# ggsave("5.figures/III. software_diffAbdundant_bar_KvS.pdf", plot, width =15)
```

```{r}
library(UpSetR)
library(ComplexHeatmap)

# make input dataframe for upsetR plot
# present of a unique phyla in each software into binary matrix
upsetR_species_matrix <- combined_DA_species%>% dplyr::select(c(Domain, Phylum,Genus,Species,software)) %>% distinct() %>% mutate(species_present = as.integer(1)) %>%pivot_wider(names_from = software, values_from = species_present) 
upsetR_species_matrix[is.na(upsetR_species_matrix)] <- 0
upsetR_species_matrix <- upsetR_species_matrix[,c("Species",softwares[softwares != "metaphlan3"], "Domain")] # reset column order
upsetR_species_matrix <- as.data.frame(upsetR_species_matrix) # erase groups and other class objects on the variable

# number of genus within each unique phylum annotation for each software
phylum_count_diffAbd <- combined_DA_species %>% group_by(software) %>% summarise(Num.Phyla = n_distinct(Phylum))

# function to check the number of phylum in each domain (ex. d__Viruses), use for query
which_domain <- function(row, domain) {
    data <- (row["Domain"]  == domain)
}

# pdf(file="5.figures/III.software_species_upset_LvK.pdf", width =9, height=5)
# pdf(file="5.figures/III.software_species_upset_LvS.pdf", width =9, height=5)
# pdf(file="5.figures/III.software_species_upset_KvS.pdf", width =9, height=5)
upset(upsetR_species_matrix,  number.angles =0,query.legend = "top",sets = c(softwares[softwares != "metaphlan3"]), 
order.by = "freq", sets.x.label = "Num.Species",set.metadata = list(data = phylum_count_diffAbd, plots = list(list(type = "hist", 
    column = "Num.Phyla", assign = 20))), queries = list(list(query = which_domain, params = list("d__Viruses"), color = "gray", active = T,query.name = "Viruses")))
dev.off()

upsetR_species_matrix %>% subset(Domain == "d__Viruses")
```

```{r}
library(pheatmap)
# check the exact phyla intersect between software using heatmap
upsetR_phylum_matrix <- combined_DA_species%>% dplyr::select(c(Domain, Phylum ,software)) %>% distinct() %>% mutate(phyla_present = as.integer(1))%>%pivot_wider(names_from = software, values_from = phyla_present) 
upsetR_phylum_matrix[is.na(upsetR_phylum_matrix)] <- 0
phyla_matrix <- data.frame(upsetR_phylum_matrix[upsetR_phylum_matrix["Phylum"]!="",])
rownames(phyla_matrix) <- phyla_matrix$Phylum
phyla_matrix_final <- as.matrix(phyla_matrix %>% dplyr::select(softwares[softwares != "metaphlan3"]))

metadata_phyla <- phyla_matrix %>% dplyr::select(Domain, Phylum) %>% dplyr::select(Domain)
domain_colors <- list(Domain = c(d__Bacteria = "#0b7fab", d__Viruses = "#e9723d", d__Archaea = "#f4d75e"))

# pdf(file="5.figures/III.software_phyla_heatmap_LvK.pdf", width=7, height=7)
# pdf(file="5.figures/III.software_phyla_heatmap_LvS.pdf", width=7, height=7)
# pdf(file="5.figures/III.software_phyla_heatmap_KvS.pdf", width=7, height=7)
pheatmap(phyla_matrix_final,col = colorRampPalette(c("lightblue", "red"))(2), annotation_row  = metadata_phyla, annotation_colors = domain_colors,  treeheight_row =0, treeheight_col = 30, legend = FALSE, fontsize_col=12, fontsize_row=10)

dev.off()
```


# higher level analyese

### aggregate reads at genus level
```{r}

for(software in names(software_list)){
  print(eval(software))
  software_obj <- paste(software,"sp", sep = "_")
  genus_obj <- tax_glom(get(software_obj), "Genus")
  assign(paste(software,"genus", sep = "_"), genus_obj)
}

genus_list <- list(blastn=blastn_genus, diamond =diamond_genus, kraken2=kraken2_genus,bracken=bracken_genus,centrifuge= centrifuge_genus,clark = clark_genus,clarks = clarks_genus,metaphlan = metaphlan3_genus,kaiju =kaiju_genus) 
# phylum taxa count: 52, 12, 39, 21, 42, 39, 34, 5, 59
# genus taxa count: 767, 79, 1058, 260, 1314, 698, 622, 13, 1614
```


### combine each software's phylum count into one dataframe
```{r}
read_genus_to_df <- function(software){
  software_obj <- paste(software, "_genus", sep = "")
  otu_df <- as.data.frame(otu_table(get(software_obj))) %>% rownames_to_column()
  otu_df_longer <- pivot_longer(otu_df, names_to = "samples", values_to = "count", cols = starts_with("R2"))
  tax_df <-as.data.frame(tax_table(get(software_obj)))%>% rownames_to_column() %>% dplyr::select(c(rowname, Genus))
  df_combined <- left_join(otu_df_longer, tax_df, by="rowname")%>% dplyr::select(!rowname)
  # df_combined <- df_combined %>% group_by(samples) %>% mutate(Total = sum(count)) %>% mutate(percent = (count/Total)*100) %>% pivot_longer(names_to = "stat", values_to = "value", cols = c(count, percent)) %>% dplyr::select(!Total)
  colnames(df_combined) <- c( "samples",  software,"Genus")
 # sum(df_combined[df_combined["samples"] == "R26.S"  & df_combined["stat"] == "percent","value"])
  df_combined
}

# read_genus_to_df("kraken2")
# genus_df[[1]]
genus_df <- lapply(softwares, read_genus_to_df)
genus_df_combined <- Reduce(function(...) merge(..., by=c("samples","Genus"),all=TRUE), genus_df)

# any_dup<- genus_df_combined %>% subset(stat =="count") %>% dplyr::select(c(samples,Genus,stat))
# any_dup[duplicated(any_dup),]
# test <- data.frame(otu_table(metaphlan3_genus))

genus_df_combined[is.na(genus_df_combined)] <- 0

genus_df_combined_longer <- genus_df_combined %>% pivot_longer(names_to = "software", values_to = "values", cols = softwares)


# some tax with same genus, but different families or order (higher ranks) cause duplication in rows, we will take the average of these genus for uniqueness
genus_df_combined_longer <- genus_df_combined_longer %>% group_by(samples,Genus,software) %>% summarise(values=mean(values))

genus_df_combined_longer %>% 
```
### keep top 10 genus for each software, combine rest phylum into one category
```{r}
top_5_genus_df <- genus_df_combined_longer %>% ungroup()%>% group_by(software,samples) %>%slice_max(order_by = values, n=3, with_ties = FALSE)


top_5_genus<- unique(top_5_genus_df$Genus)

rest_genus_df <- genus_df_combined_longer %>% subset(!Genus %in% top_5_genus) %>% group_by(samples, software) %>% summarise(values =sum(values)) %>% mutate(Genus = "g__Other.Genus")

genus_df_simplified <- rbind(top_5_genus_df, rest_genus_df)
unique(genus_df_simplified$Genus)
genus_df_simplified$Genus <- factor(genus_df_simplified$Genus, levels= c("g__Other.Genus", unique(genus_df_simplified$Genus)[-49]))
```
### plot genus level abundance
```{r}

manual_color <-c("#A0C4FF","#80B918","#FEE440","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#9BF6FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C")

genus_df_simplified$software <- factor(genus_df_simplified$software, levels = softwares)
# genus_df_simplified$stat <- factor(genus_df_simplified$stat , levels =c("count", "percent"))
p1 <- ggplot(genus_df_simplified , aes(x= samples, y= values,fill=Genus))+
  geom_bar(stat = "identity")+
  facet_wrap(~software, scale="free_y", ncol = 1, strip.position = "left")+
    theme_classic()+
  theme(text=element_text(size=20), title=element_text(size=17),axis.text.x = element_text(angle=90,hjust=0.95,vjust=0.2),tagger.panel.tag.background = element_rect(linetype = 0,fill = alpha("white", 0)),legend.position = "none",panel.spacing  = unit(0.08, "cm"), axis.title.x = element_blank())+
  tag_facets()+
guides(fill=guide_legend(ncol=2, bycol=TRUE))+
  scale_y_continuous(label=comma)+
  labs(y="genus abudance count")+
  scale_fill_manual(values = manual_color)
p1
p2 <- ggplot(genus_df_simplified , aes(x= samples, y= values,fill=Genus))+
  geom_bar(stat = "identity", position = "fill")+
  facet_wrap(~software, scale="free_y", ncol = 1)+
    theme_classic()+
  theme(text=element_text(size=20), title=element_text(size=17),axis.text.x = element_text(angle=90,hjust=0.95,vjust=0.2),tagger.panel.tag.background = element_rect(linetype = 0,fill = alpha("white", 0)),strip.background = element_blank(), strip.text = element_blank(),legend.position ="none", axis.title.y.right  = element_text(margin = margin(l = 10)), panel.spacing = unit(0.08, "cm"), axis.title.x = element_blank())+
  # tag_facets()+
guides(fill=guide_legend(ncol=2, bycol=TRUE))+
  scale_y_continuous(label=comma,position = "right")+
  labs(y="genus abudance count")+
  scale_fill_manual(values = manual_color)

pc <- grid.arrange(p1,p2,ncol=2)
ggsave("5.figures/II.software_genus_bar.pdf",pc,width =15, height=15)

unique(genus_df_simplified$Genus)
l <- ggplot(genus_df_simplified , aes(x= samples, y= values,fill=Genus))+
  geom_bar(stat = "identity")+
  facet_wrap(~software, scale="free_y", ncol = 1)+
    theme_classic()+
  theme(text=element_text(size=20), title=element_text(size=17),axis.text.x = element_text(angle=90,hjust=0.95,vjust=0.2),tagger.panel.tag.background = element_rect(linetype = 0,fill = alpha("white", 0)),strip.background = element_blank(), strip.text = element_blank(), axis.title.y.right  = element_text(margin = margin(l = 10)), panel.spacing = unit(0.08, "cm"), axis.title.x = element_blank())+
  # tag_facets()+
guides(fill=guide_legend(ncol=1, bycol=TRUE))+
  scale_y_continuous(label=comma,position = "right")+
  labs(y="genus abudance count")+
  scale_fill_manual(values = manual_color)
# ggsave("5.figures/II.software_genus_bar.legend.pdf",l,width =15, height=15)

```

### phylum level bar plot
```{r}
softwar <- "kaiju"
read_phylum_to_df <- function(software){
  software_obj <- paste(software, "_sp", sep = "") # we will use the genus level object for phylum aggregation
  otu_df <- as.data.frame(otu_table(get(software_obj))) %>% rownames_to_column()
  otu_df_longer <- pivot_longer(otu_df, names_to = "samples", values_to = "count", cols = starts_with("R2"))
  tax_df <-as.data.frame(tax_table(get(software_obj)))%>% rownames_to_column() %>% dplyr::select(c(rowname, Domain,Phylum, Genus, Species))
  df_combined <- left_join(otu_df_longer, tax_df, by="rowname")%>% dplyr::select(!rowname) %>% group_by(samples, Domain,Phylum, Genus, Species) %>% summarise(count=sum(count))
  
  colnames(df_combined) <- c( "samples",  "Domain", "Phylum","Genus","Species",software)
 # sum(df_combined[df_combined["samples"] == "R26.S"  & df_combined["stat"] == "percent","value"])
  df_combined
}

# read_phylum_to_df("kraken2")

phylum_df <- lapply(softwares, read_phylum_to_df)
phylum_df[[1]] %>% subset(Phylum == "p__Firmicutes") %>% subset(blastn!=0) %>% arrange(desc(blastn)) %>% subset(samples == "R27.S") %>% summarise(unique(Genus))
phylum_df[[2]] %>% subset(Phylum == "p__Firmicutes") %>% subset(diamond!=0) %>% arrange(desc(diamond)) %>% subset(samples == "R28.L") #%>% summarise(unique(Genus))
phylum_df[[3]] %>% subset(Phylum == "p__Firmicutes") %>% subset(kraken2!=0) %>% arrange(desc(kraken2)) %>% subset(samples == "R27.S")%>% summarise(unique(Genus))
phylum_df[[4]] %>% subset(Phylum == "p__Firmicutes") %>% subset(bracken!=0) %>% arrange(desc(bracken)) %>% subset(samples == "R27.S")%>% summarise(unique(Genus))
phylum_df[[5]] %>% subset(Phylum == "p__Firmicutes") %>% subset(centrifuge!=0) %>% arrange(desc(centrifuge)) %>% subset(samples == "R27.S")#%>% summarise(unique(Genus))
phylum_df[[6]] %>% subset(Phylum == "p__Firmicutes") %>% subset(clark!=0) %>% arrange(desc(clark)) %>% subset(samples == "R27.L")%>% summarise(unique(Genus))
phylum_df[[7]] %>% subset(Phylum == "p__Firmicutes") %>% subset(clarks!=0) %>% arrange(desc(clarks)) %>% subset(samples == "R27.L")%>% summarise(unique(Genus))
phylum_df[[8]] %>% subset(Phylum == "p__Firmicutes") %>% subset(metaphlan3!=0) %>% arrange(desc(metaphlan3)) %>% subset(samples == "R27.L")%>% summarise(unique(Genus))#%>% group_by(Domain,Phylum, Genus, Species) %>% summarise(sum(metaphlan3))
phylum_df[[9]] %>% subset(Phylum == "p__Firmicutes") %>% subset(kaiju!=0) %>% arrange(desc(kaiju)) %>% subset(samples == "R27.L")%>% summarise(unique(Genus))


phylum_df_combined <- Reduce(function(...) merge(..., by=c("samples","Phylum"),all=TRUE), phylum_df)

# any_dup<- genus_df_combined %>% subset(stat =="count") %>% dplyr::select(c(samples,Genus,stat))
# any_dup[duplicated(any_dup),]
# test <- data.frame(otu_table(metaphlan3_genus))

phylum_df_combined[is.na(phylum_df_combined)] <- 0

phylum_df_combined_longer <- phylum_df_combined %>% pivot_longer(names_to = "software", values_to = "values", cols = softwares)


# # some tax with same genus, but different families or order (higher ranks) cause duplication in rows, we will take the average of these genus for uniqueness
# phylum_df_combined_longer <- phylum_df_combined_longer %>% group_by(samples,Phylum,software) %>% summarise(values=mean(values))

```

### keep top 10 phylum for each software, combine rest phylum into one category
```{r}
top_5_phyla_df <- phylum_df_combined_longer %>% ungroup()%>% group_by(software,samples) %>%slice_max(order_by = values, n=5, with_ties = FALSE) %>% mutate(Phylum = sapply(Phylum, function(x){ifelse(x != "", unlist(strsplit(x, " ", fixed = TRUE))[1],"")})) %>% subset(Phylum != "")


top_5_phyla<- unique(top_5_phyla_df[top_5_phyla_df["Phylum"] != "",]$Phylum)

rest_phylum_df <- phylum_df_combined_longer %>% subset(!Phylum %in% top_5_phyla) %>% mutate(Phylum = sapply(Phylum, function(x){
  s <- unlist(strsplit(x, " ", fixed = TRUE))[1]
  ifelse(x=="", "p__", s)
}))%>% group_by(samples, software) %>% summarise(values =sum(values)) %>% mutate(Phylum = "p__Other.Phyla")


phyla_df_simplified <- rbind(top_5_phyla_df, rest_phylum_df)


phyla_df_simplified$Phylum <- factor(phyla_df_simplified$Phylum, levels= c("p__Other.Phyla", unique(phyla_df_simplified$Phylum)[-22]))

phylum_df_combined_longer %>% subset(samples == "R27.L" & software == "diamond" &values != 0)  %>% group_by(software) %>% summarise(n_distinct(Phylum))
```

### plot phylum level abundance
```{r}

manual_color <-c("#A0C4FF","#80B918","#FFC6FF","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#9BF6FF","#BDB2FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#AD8EB0ff","#F79D84","#cccccc","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#FEE440","#54478C")

# genus_df_simplified$stat <- factor(genus_df_simplified$stat , levels =c("count", "percent"))
phyla_df_simplified$software<- factor(phyla_df_simplified$software, levels = softwares)
p1 <- ggplot(phyla_df_simplified , aes(x= samples, y= values,fill=Phylum))+
  geom_bar(stat = "identity")+
  facet_wrap(~software, scale="free_y", ncol = 1, strip.position = "left")+
    theme_classic()+
  theme(text=element_text(size=20), title=element_text(size=17),axis.text.x = element_text(angle=90,hjust=0.95,vjust=0.2),tagger.panel.tag.background = element_rect(linetype = 0,fill = alpha("white", 0)),legend.position = "none",panel.spacing  = unit(0.08, "cm"), axis.title.x = element_blank())+
  tag_facets()+
guides(fill=guide_legend(ncol=2, bycol=TRUE))+
  scale_y_continuous(label=comma)+
  labs(y="genus abudance count")+
  scale_fill_manual(values = manual_color)
p1
p2 <- ggplot(phyla_df_simplified , aes(x= samples, y= values,fill=Phylum))+
  geom_bar(stat = "identity", position = "fill")+
  facet_wrap(~software, scale="free_y", ncol = 1)+
    theme_classic()+
  theme(text=element_text(size=20), title=element_text(size=17),axis.text.x = element_text(angle=90,hjust=0.95,vjust=0.2),tagger.panel.tag.background = element_rect(linetype = 0,fill = alpha("white", 0)),strip.background = element_blank(), strip.text = element_blank(),legend.position ="none", axis.title.y.right  = element_text(margin = margin(l = 10)), panel.spacing = unit(0.08, "cm"), axis.title.x = element_blank())+
  # tag_facets()+
guides(fill=guide_legend(ncol=2, bycol=TRUE))+
  scale_y_continuous(label=comma,position = "right")+
  labs(y="genus abudance count")+
  scale_fill_manual(values = manual_color)
p2
pc <- grid.arrange(p1,p2,ncol=2)
# ggsave("5.figures/II.software_phylum_bar.pdf",pc,width =15, height=15)

l <- ggplot(phyla_df_simplified , aes(x= samples, y= values,fill=Phylum))+
  geom_bar(stat = "identity")+
  facet_wrap(~software, scale="free_y", ncol = 1)+
    theme_classic()+
  theme(text=element_text(size=20), title=element_text(size=17),axis.text.x = element_text(angle=90,hjust=0.95,vjust=0.2),tagger.panel.tag.background = element_rect(linetype = 0,fill = alpha("white", 0)),strip.background = element_blank(), strip.text = element_blank(), axis.title.y.right  = element_text(margin = margin(l = 10)), panel.spacing = unit(0.08, "cm"), axis.title.x = element_blank())+
  # tag_facets()+
guides(fill=guide_legend(ncol=1, bycol=TRUE))+
  scale_y_continuous(label=comma,position = "right")+
  labs(y="genus abudance count")+
  scale_fill_manual(values = manual_color)
# ggsave("5.figures/II.software_phylum_bar.legend.pdf",l,width =15, height=15)

t1 <- phyla_df_simplified %>% subset(samples %in% c("R27.L") & Phylum == "p__Proteobacteria")
t2 <- phyla_df_simplified %>% subset(samples %in%c("R27.L") ) %>% group_by(software, samples) %>% summarise(total=sum(values))#%>% summarise(total = sum(values))
t3 <- left_join(t2,t1, by = c("samples", "software"))
t3$percent <- t3$values/t3$total
t3 %>% subset(software != "metaphlan3") %>% ungroup()%>% summarise(mean = mean(percent), sd = sd(percent))
t4 <- t3 %>% dplyr::select(-Phylum)
t4[is.na(t4)] <- 0
t4 %>% subset(software != "diamond" & software != "centrifuge") %>% summarise(mean=mean(percent), sd = sd(percent))
phyla_df_simplified %>% subset(Phylum == "p__Pisuviricota")
phyla_df_simplified %>%  subset(Phylum == "p__Firmicutes")#%>% summarise(total = sum(values))
phyla_df_simplified %>% subset(software =="metaphlan3")
```


```{r}

# openxlsx::saveWorkbook(software.xlsx, file = "4.tables/II.software_comparison_full.xlsx", overwrite = TRUE)
```